cloudinit/sources/DataSourceWSL.py

```diff
@@ -237,23 +268,84 @@ def check_instance_id(self, sys_cfg) -> bool:
             return False
 
     def _get_data(self) -> bool:
-        self.vendordata_raw = None
-        seed_dir = cloud_init_data_dir()
+        if not subp.which(WSLPATH_CMD):
+            LOG.debug(
+                "No WSL command %s found. Cannot detect WSL datasource",
+                WSLPATH_CMD,
+            )
+            return False
+        self.instance_name = instance_name()
+
+        try:
+            user_home = find_home()
+        except IOError as e:
+            LOG.debug("Unable to detect WSL datasource: %s", e)
+            return False
 
+        seed_dir = cloud_init_data_dir(user_home)
+        user_data: Optional[Union[dict, bytes]] = None
+
+        # Load any metadata
         try:
             self.metadata = load_instance_metadata(
                 seed_dir, self.instance_name
             )
-            file = self.find_user_data_file(seed_dir)
-            self.userdata_raw = cast(
-                str, util.load_binary_file(file.as_posix())
-            )
-            return True
+        except (ValueError, IOError) as err:
+            LOG.error("Unable to load metadata: %s", str(err))
+            return False
+
+        # Load Ubuntu Pro configs
+        agent_data, user_data = load_ubuntu_pro_data(user_home)
 
+        # Load regular user configs
+        try:
+            if user_data is None and seed_dir is not None:
+                file = self.find_user_data_file(seed_dir)
+                user_data = load_yaml_or_bin(file.as_posix())
         except (ValueError, IOError) as err:
-            LOG.error("Unable to setup WSL datasource: %s", str(err))
+            LOG.error(
+                "Unable to load any user-data file in %s: %s",
+                seed_dir,
+                str(err),
+            )
+
+        # No configs were found
+        if not any([user_data, agent_data]):
             return False
 
+        # If we cannot reliably model data files as dicts, then we cannot merge
+        # ourselves, so we can pass the data in ascending order as a list for
+        # cloud-init to handle internally
+        if isinstance(agent_data, bytes) or isinstance(user_data, bytes):
```

Comment:
From cloud-init's point of view, is it better to be strict with what we know from the UP4W implementation (which our team controls and do enforce YAML as of now) or let the datasource be more flexible so that changes in UP4W implementation won't affect the datasource as much?

The question sounds silly in principle, but there is a layer of trust here: if we know the agent only writes YAML, ingesting any other format would mean ingesting invalid (thus potentially harmful) config data. I'm in favor of the less future maintenance path, trusting that the user has full control over the directories involved and apply the relevant protection to their systems. I hope I'm not too positive about this 😄